File: src/generator/set.ts
import { ArbiContainer, Generator } from '../Generator'
import { Shrinkable } from '../Shrinkable'
import { shrinkableSet } from '../shrinker/set'

const DEFAULT_MIN_SIZE = 0
const DEFAULT_MAX_SIZE = 20

/**
 * Configuration for {@link SetGen}.
 *
 * @example
 * ```ts
 * Gen.set({ elemGen: Gen.asciiString(1, 3), maxSize: 5 })
 * ```
 */
export interface SetGenConfig<T> {
    /** Generator for the elements to be included in the set. Required. */
    elemGen: Generator<T>
    /** Minimum number of elements. Default: 0. */
    minSize?: number
    /** Maximum number of elements. Default: 20. */
    maxSize?: number
}

/**
 * Creates a Generator for producing Set<T> instances.
 * Supports both a positional form and a config-object form.
 *
 * @example
 * ```ts
 * // positional (existing)
 * Gen.set(Gen.asciiString(1, 3), 0, 5)
 * // config object — all size fields optional
 * Gen.set({ elemGen: Gen.asciiString(1, 3), maxSize: 5 })
 * ```
 */
export function SetGen<T>(config: SetGenConfig<T>): Generator<Set<T>>
export function SetGen<T>(elemGen: Generator<T>, minSize: number, maxSize: number): Generator<Set<T>>
export function SetGen<T>(
    first: SetGenConfig<T> | Generator<T>,
    minSize?: number,
    maxSize?: number
): Generator<Set<T>> {
    const isPosForm = minSize !== undefined
    const resolvedElemGen = isPosForm ? (first as Generator<T>) : (first as SetGenConfig<T>).elemGen
    const resolvedMinSize = isPosForm ? minSize! : ((first as SetGenConfig<T>).minSize ?? DEFAULT_MIN_SIZE)
    const resolvedMaxSize = isPosForm ? maxSize! : ((first as SetGenConfig<T>).maxSize ?? DEFAULT_MAX_SIZE)

    return new ArbiContainer<Set<T>>(
        rand => {
            // Determine the target size for the set randomly within the specified range.
            const size = rand.interval(resolvedMinSize, resolvedMaxSize)
            const array: Array<Shrinkable<T>> = []
            const valueSet: Set<T> = new Set([])
            // Keep generating elements until the set reaches the target size.
            // Ensures uniqueness by checking if the value already exists in valueSet.
            while (array.length < size) {
                const shr = resolvedElemGen.generate(rand)
                if (!valueSet.has(shr.value)) {
                    array.push(shr)
                    valueSet.add(shr.value)
                }
            }
            // Create a shrinkable set from the generated unique elements.
            return shrinkableSet(array, resolvedMinSize)
        },
        resolvedMinSize,
        resolvedMaxSize
    )
}

File: dartproptest/lib/src/generator/set.dart
import '../generator.dart';
import '../shrinkable.dart';
import '../shrinker/set.dart';

/// Creates a Generator for producing Set<T> instances.
///
/// [elemGen] The generator for the elements to be included in the set.
/// [minSize] The minimum number of elements the generated set should contain.
/// [maxSize] The maximum number of elements the generated set should contain.
/// Returns a Generator that produces Set<T> instances with sizes between minSize and maxSize (inclusive).
Generator<Set<T>> setGen<T>(Generator<T> elemGen, int minSize, int maxSize) {
  return Arbitrary<Set<T>>((rand) {
    // Determine the target size for the set randomly within the specified range.
    final size = rand.interval(minSize, maxSize);
    final array = <Shrinkable<T>>[];
    final valueSet = <T>{};

    // Keep generating elements until the set reaches the target size.
    // Ensures uniqueness by checking if the value already exists in valueSet.
    int attempts = 0;
    const maxAttempts = 1000; // Prevent infinite loops
    while (array.length < size && attempts < maxAttempts) {
      final shr = elemGen.generate(rand);
      if (!valueSet.contains(shr.value)) {
        array.add(shr);
        valueSet.add(shr.value);
      }
      attempts++;
    }

    // If we couldn't generate enough unique elements, use what we have
    if (array.length < size) {
      // This shouldn't happen in normal cases, but provides a safety net
      // We'll use the elements we managed to generate
    }

    // Create a shrinkable set from the generated unique elements.
    return shrinkableSet(array, minSize);
  });
}
